api/grpcserver/transaction_service.go

```diff
@@ -226,42 +226,48 @@ func (s TransactionService) TransactionsStateStream(in *pb.TransactionsStateStre
 				return nil
 			}
 			// Filter for any matching transactions in the reported layer
-			// TODO: this is inefficient and could be optimized!
-			// See https://github.com/spacemeshos/go-spacemesh/issues/2076
 			for _, b := range layer.Layer.Blocks() {
-				for _, layerTxid := range b.TxIDs {
-					for _, txid := range in.TransactionId {
-						if bytes.Equal(layerTxid.Bytes(), txid.Id) {
-							var txstate pb.TransactionState_TransactionState
-							switch layer.Status {
-							case events.LayerStatusTypeApproved:
-								txstate = pb.TransactionState_TRANSACTION_STATE_MESH
-							case events.LayerStatusTypeConfirmed:
-								txstate = pb.TransactionState_TRANSACTION_STATE_PROCESSED
-							default:
-								txstate = pb.TransactionState_TRANSACTION_STATE_UNSPECIFIED
-							}
-							res := &pb.TransactionsStateStreamResponse{
-								TransactionState: &pb.TransactionState{
-									Id:    txid,
-									State: txstate,
-								},
-							}
-							if in.IncludeTransactions {
-								tx, err := s.Mesh.GetTransaction(layerTxid)
-								if err != nil {
-									log.Error("could not find transaction %v from layer %v: %v", layerTxid, layer, err)
-									return status.Error(codes.Internal, "error retrieving tx data")
-								}
+				blockTXIDSet := make(map[types.TransactionID]struct{})
 
-								res.Transaction = convertTransaction(tx)
-							}
-							if err := stream.Send(res); err != nil {
-								return err
+				//create a set for the block transaction IDs
+				for _, txid := range b.TxIDs {
+					blockTXIDSet[txid] = struct{}{}
+				}
+
+				for _, inputTxID := range in.TransactionId {
+					var arrayID [32]byte
+					copy(arrayID[:], inputTxID.Id[:])
```

Comment:
This copy seems unnecessary

Reply:
I can only convert it into a TransactionID if it's a fixed size array, and so I created a fixed sized array from the slice

Reply:
You're right. The underlying issue here is that data items including txid do not have fixed lengths in our protobuf API. I opened https://github.com/spacemeshos/api/issues/130 to investigate if there's anything we can do about this.

Could you please add an explanatory comment here? Something to the effect of, "Since the txid coming in from the API does not have a fixed length, we need to convert it from a slice into an array here before we can convert it into a TransactionID object. We don't need to worry about error-handling here since `copy()` intelligently copies only as much as it can. If the resultant TransactionID is invalid, an error will be returned below."